File: packages/wasmvm/wasmlib/src/wasmtypes/sctokenid.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_TOKEN_ID_LENGTH: usize = 50;

#[derive(PartialEq, Clone, Copy, Eq, Hash)]
pub struct ScTokenID {
    id: [u8; SC_TOKEN_ID_LENGTH],
}

impl ScTokenID {
    pub fn to_bytes(&self) -> Vec<u8> {
        token_id_to_bytes(self)
    }

    pub fn to_string(&self) -> String {
        token_id_to_string(self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn token_id_decode(dec: &mut WasmDecoder) -> ScTokenID {
    let buf = dec.fixed_bytes(SC_TOKEN_ID_LENGTH);
    ScTokenID { id: buf.try_into().expect("WTF?") }
}

pub fn token_id_encode(enc: &mut WasmEncoder, value: &ScTokenID) {
    enc.fixed_bytes(&value.id, SC_TOKEN_ID_LENGTH);
}

pub fn token_id_from_bytes(buf: &[u8]) -> ScTokenID {
    if buf.len() == 0 {
        return ScTokenID { id: [0; SC_TOKEN_ID_LENGTH] };
    }
    if buf.len() != SC_TOKEN_ID_LENGTH {
        panic("invalid TokenID length");
    }
    ScTokenID { id: buf.try_into().expect("WTF?") }
}

pub fn token_id_to_bytes(value: &ScTokenID) -> Vec<u8> {
    value.id.to_vec()
}

pub fn token_id_to_string(value: &ScTokenID) -> String {
    // TODO standardize human readable string
    base58_encode(&value.id)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableTokenID {
    proxy: Proxy,
}

impl ScImmutableTokenID {
    pub fn new(proxy: Proxy) -> ScImmutableTokenID {
        ScImmutableTokenID { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        token_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScTokenID {
        token_id_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScTokenID in host container
pub struct ScMutableTokenID {
    proxy: Proxy,
}

impl ScMutableTokenID {
    pub fn new(proxy: Proxy) -> ScMutableTokenID {
        ScMutableTokenID { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &ScTokenID) {
        self.proxy.set(&token_id_to_bytes(&value));
    }

    pub fn to_string(&self) -> String {
        token_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScTokenID {
        token_id_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/sctokenid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from '../sandbox';
import {hexDecode, hexEncode, WasmDecoder, WasmEncoder, zeroes} from './codec';
import {Proxy} from './proxy';
import {bytesCompare} from './scbytes';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScTokenIDLength = 38;

export class ScTokenID {
    id: Uint8Array = zeroes(ScTokenIDLength);

    public equals(other: ScTokenID): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return tokenIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return tokenIDToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function tokenIDDecode(dec: WasmDecoder): ScTokenID {
    return tokenIDFromBytesUnchecked(dec.fixedBytes(ScTokenIDLength));
}

export function tokenIDEncode(enc: WasmEncoder, value: ScTokenID): void {
    enc.fixedBytes(value.id, ScTokenIDLength);
}

export function tokenIDFromBytes(buf: Uint8Array | null): ScTokenID {
    if (buf === null || buf.length == 0) {
        return new ScTokenID();
    }
    if (buf.length != ScTokenIDLength) {
        panic('invalid TokenID length');
    }
    return tokenIDFromBytesUnchecked(buf);
}

export function tokenIDToBytes(value: ScTokenID): Uint8Array {
    return value.id;
}

export function tokenIDFromString(value: string): ScTokenID {
    return tokenIDFromBytes(hexDecode(value));
}

export function tokenIDToString(value: ScTokenID): string {
    return hexEncode(value.id);
}

function tokenIDFromBytesUnchecked(buf: Uint8Array): ScTokenID {
    const o = new ScTokenID();
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableTokenID {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return tokenIDToString(this.value());
    }

    value(): ScTokenID {
        return tokenIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableTokenID extends ScImmutableTokenID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScTokenID): void {
        this.proxy.set(tokenIDToBytes(value));
    }
}
